File: src/css_trans.f
!                     ********************
                      SUBROUTINE CSS_TRANS
!                     ********************
     & (AV,AC,AR,NZ,NR,NC,TV,TC,TR)
!
!***********************************************************************
! CSS PACKAGE - SERGIO CASTIBLANCO
!***********************************************************************
!
!brief    1) COMPUTES THE TRANSPOSE OF MATRIX A IN CSS (CSR or CSC)
!
!history  Sergio Castiblanco
!+        21/02/2022
!+        Translation for original Matlab implementation
!
!+
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| AV       |-->| VALUES OF MATRIX A IN CSS                            |
!| AC       |-->| ROW INDICES IF CSC, COLUMN INDICES IF CSR            |
!| AR       |-->| COLUMN INDICES IF CSC, ROW INDICES IF CSR            |
!| NZ       |-->| NUMBER OF NON-ZEROS ENTRIES OF A                     |
!| NR       |-->| NUMBER OF ROWS OF A (CSR), NUMBER OF COLUMNS (CSC)   |
!| NC       |-->| NUMBER OF COLUMNS OF A(CSR), NUMBER OF ROWS(CSC)     |
!| TV       |<--| VALUES OF MATRIX A^T IN CSS                          |
!| TC       |<--| RESPECTIVE ROW OR COLUMN INDICES                     |
!| TR       |<--| RESPECTIVE COLUMN OR ROW INDICES                     |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE CSC_STORAGE, EX_CSS_TRANS => CSS_TRANS
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      INTEGER, INTENT(IN) :: NZ,NR,NC
      DOUBLE PRECISION, DIMENSION(NZ), INTENT(IN) :: AV
      INTEGER, DIMENSION(NZ), INTENT(IN) :: AC
      INTEGER, DIMENSION(NR+1), INTENT(IN) :: AR
      DOUBLE PRECISION, DIMENSION(NZ), INTENT(OUT) :: TV
      INTEGER, DIMENSION(NZ), INTENT(OUT) :: TC
      INTEGER, DIMENSION(NC+1), INTENT(OUT) :: TR
!
!  IN SUBROUTINE VARIABLES
!
      INTEGER :: MH,I,J,K,JP,IAA,IAB
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      MH = NC + 1
!
      DO I=1,NZ
        J = AC(I)+2
        IF(J.LE.MH)THEN
          TR(J) = TR(J)+1
        ENDIF
      ENDDO
!
      TR(1) = AR(1)
      TR(2) = AR(1)
!
      DO I=3,MH
        TR(I) = TR(I) + TR(I-1)
      ENDDO
!
      DO I=1,NR
        IAA = AR(I)
        IAB = AR(I+1)-AR(1)
        IF(IAB.LT.IAA)THEN
          EXIT
        ENDIF
        DO JP=IAA,IAB
          J = AC(JP) + 1
          K = TR(J)
          TC(K) = I
          TV(K) = AV(JP)
          TR(J) = K+1
        ENDDO
      ENDDO
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE CSS_TRANS

File: matlab/css_trans.m
function [Tv,Tc,Tr] = css_trans(Av,Ac,Ar)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% This function takes the matrix A in CS storage, and computes its
% transpose such as:
%
%      T = A'
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Number of columns and non-zeros of A
n = length(Ar)-1;
nz = Ar(n+1)-1;

%Number of columns of T
mt = max(Ac);

%Allocating Tv, Tr and Tc
Tv = zeros(1,nz);
Tc = zeros(1,nz);
Tr = zeros(1,mt+1);

mh = mt+1;

for i=1:nz
    j = Ac(i)+2;
    if(j<=mh)
        Tr(j)=Tr(j)+1;
    end
end

Tr(1) = 1;
Tr(2) = 1;

for i=3:mh
    Tr(i)=Tr(i)+Tr(i-1);
end

for i=1:n
    iaa = Ar(i);
    iab = Ar(i+1)-1;
    if iab<iaa
        break
    end
    for jp=iaa:iab
        j=Ac(jp)+1;
        k=Tr(j);
        Tc(k)=i;
        Tv(k)=Av(jp);
        Tr(j)=k+1;
    end
end

end